crates/flowctl/src/catalog/publish.rs

```diff
@@ -0,0 +1,73 @@
+use crate::{controlplane, draft, source, CliContext};
+use anyhow::Context;
+
+#[derive(Debug, clap::Args)]
+pub struct Publish {
+    #[clap(flatten)]
+    pub source_args: source::SourceArgs,
+    /// Proceed with the publication without prompting for confirmation.
+    ///
+    /// Normally, publish will stop and ask for confirmation before it proceeds. This disables that confirmation.
+    /// This flag is required if running flowctl non-interactively, such as in a shell script.
+    #[clap(long)]
+    pub auto_approve: bool,
+}
+
+pub async fn do_publish(ctx: &mut CliContext, args: &Publish) -> anyhow::Result<()> {
+    use crossterm::tty::IsTty;
+
+    // The order here is intentional, to minimize the number of things that might be left dangling
+    // in common error scenarios. For example, we don't create the draft until after bundling, because
+    // then we'd have to clean up the empty draft if the bundling fails. The very first thing is to create the client,
+    // since that can fail due to missing/expired credentials.
+    let client = ctx.controlplane_client()?;
+
+    anyhow::ensure!(args.auto_approve || std::io::stdin().is_tty(), "The publish command must be run interactively unless the `--auto-approve` flag is provided");
+
+    let sources = args.source_args.resolve_sources().await?;
+    let catalog = crate::source::bundle(sources).await?;
+
+    let draft = draft::create_draft(client.clone()).await?;
+    println!("Created draft: {}", &draft.id);
```

Comment:
`println!` goes to stdout and is intended to be messages that we always want to show to the user. `tracing::info!` is just application logs, which go to stderr, and info level isn't even enabled by default. Here the `println`s are meant to communicate progress, since otherwise the user may just be waiting for a while with no output.